rdkit/Chem/Draw/IPythonConsoleIntegration.py

```diff
@@ -0,0 +1,558 @@
+    """ Handles ligand labeling (conf label and atom label) in update3D function """
+    
+    uid = molViewState.uid
+    wgData = globals()['wgData'][uid]
+    sMolNames = molViewState.selectedMolNames
+    sConfIds = molViewState.selectedConfIds
+    
+    if not len(sMolNames)==1 and len(sConfIds)==1:
+        return None
+    else:
+        if 'confLabel' in wgData:
+            if wgData['confLabel'].value:
+                confLabel = list(sMolNames)[0] + ':' + str(wgData['confId'].value)
+                view.addLabel(confLabel, {'backgroundColor':'gray', 'fontColor':'white',
+                                          'showBackground':'true', 'alignment':'bottomCenter'})
+        
+        if 'atomLabel' in wgData:
+            
+            mol = molViewState.moldict[list(sMolNames)[0]]
+            confId=wgData['confId'].value
+            sconf = mol.GetConformer(confId)
+            xyz = sconf.GetPositions()
+            
+            if wgData['atomLabel'].value:
+                for i in list(range(sconf.GetNumAtoms())):
+                    atomLabel = '{}{}'.format(mol.GetAtomWithIdx(i).GetSymbol(), i)
+                    view.addLabel(atomLabel, {'inFront' : 'false', 
+                                              'fontSize' : '12',
+                                              'fontColor':'gray',
+                                              'showBackground':'false',
+                                              'position' : {'x':xyz[i][0], 'y':xyz[i][1], 'z':xyz[i][2]}
+                                             })
+    
+    return view
+    
+def handleProteinStyle(molViewState, view, pModelNum):
+    """ Handles protein drawing style in update3D function """
+    uid = molViewState.uid
+    wgData = globals()['wgData'][uid]
+    
+    # 'pStyle_wg' and 'pStyle_helicesAsTubes_wg' rendered as pair
+    if 'pStyle_wg' in wgData:
+        pStyle=wgData['pStyle_wg'].value
+    else:
+        pStyle = REPRESENTATION_DEFAULT['protein']
+        
+    
+    if pStyle == 'surface':
+        view.addSurface('SES', {'model':pModelNum});
+    elif pStyle == 'line':
+        view.setStyle({'model':pModelNum},{'line':{}});
+    elif pStyle == 'cartoon' and 'pStyle_helicesAsTubes_wg' not in wgData:
+        view.setStyle({'model':pModelNum},{'cartoon':{'color': 'spectrum', 'arrows': 'true'}})
+    else:
+        if wgData['pStyle_helicesAsTubes_wg'].value:
+            view.setStyle({'model':pModelNum},{'cartoon':{'color': 'spectrum',
+                                                          'arrows': 'true', 'tubes' : 'true'}})
+        else:
+            view.setStyle({'model':pModelNum},{'cartoon':{'color': 'spectrum', 'arrows': 'true'}})
+    
+    return view
+    
+    
+def update3D(model_id):
+    """ This function invoked whenever user interacts with widgets.
+    It runs first time through handle_button() when the start button clicked.
+    """
+    uid = globals()['wgMIds'][model_id]
```

Comment:
This is the only reason that `globals()['wgMIds']` is require; to map the widgets model IDs to the specific viewer data. These are viewInstantiated, wgData (state of the controls?), and molViewState (state of molecule viewer?). If you bundle these three pieces of data into an object, then you only need to associated this object with the widgets. Let me try to make it clearer what I mean:
```
class MoleculeView(object):
  viewInstantiated = false
  def __init__(wgData, molViewState):
    this.viewInstantiated = false
    this.wgData = wgData
    this.molViewState = molViewState

  def update3D():
    ... you now have immediately access to the correct data ...
```
I move `update3D` into this class; think about what other functions could become methods of this class. `view` could also be a field of this class; currently it's stored in wgData, however it's not really a widget it has more to do with the MoleculeView. 

Further down, it would basically be:
```
moleculeView = MoleculeView(wgData, molViewState)
globals()['wgMIds'][model_id] = moleculeView
```
The function `handleChange` would become:
```
def handle_change(change):
  """This function handles all the interactive widgets except buttons and 3Dmol.js viewer"""
  globals()['wgMIds'][change.owner.model_id].update3D()
```
for clarity you could consider a function:
```
def getMoleculeView(model_id):
  return globals()['wgMIds'][change.owner.model_id]

def handle_change(change):
  """This function handles all the interactive widgets except buttons and 3Dmol.js viewer"""
  getMoleculeView(change.owner.model_id).update3D()
```